lib/cpp/base_model/model_generalized_linear.cpp

```diff
@@ -26,20 +40,23 @@ const char *ModelGeneralizedLinear::get_class_name() const {
   return "ModelGeneralizedLinear";
 }
 
-
-double ModelGeneralizedLinear::grad_i_factor(const ulong i,
-                                             const ArrayDouble &coeffs) {
+template <class T, class K>
+K
+TModelGeneralizedLinear<T, K>::grad_i_factor(const ulong i,
+                                             const Array<T> &coeffs) {
```

Comment:
Dummy question: is there a way to somehow "cast" Array<atomic double> into Array<double> if this array will be read-only? Because this is what will happen in most mdoels: coeffs will be read to compute gradient / loss but a model itself won't write on coeffs.